src/parser/core/SharedCode.tsx

```diff
@@ -0,0 +1,231 @@
+   * @param preCastSpell The spell object of the specific spell (or an array of spells) you want to check for
+   * @returns a true/false of whether the precast was found or not.
+   */
+  checkPreCast(event: AnyEvent, preCastSpell: SpellInfo | SpellInfo[]) {
+    const preCast = this.getPreCast(event, preCastSpell);
+    return preCast ? true : false;
+  }
+
+  /**
+   * @param event the event you want to get the pre cast for
+   * @param preCastSpell an optional spell that you want to look for. Leave undefined to get the precast spell regardless of what spell it is.
+   * @returns the event for the precast spell, or undefined if none was found.
+   */
+  getPreCast(event: AnyEvent, preCastSpell?: SpellInfo | SpellInfo[]) {
+    return this.getEvents(true, EventType.Cast, preCastSpell, 1, event.timestamp, 250)[0];
+  }
+
+  /**
+   * @param eventType The type of event you want to count. i.e. EventType.Cast, EventType.BeginCast, etc. Use EventType.Event for all events.
+   * @param spell the spell you want to count casts for.
+   * @returns the number of events found
+   */
+  countEvents(eventType: EventType, spell: SpellInfo) {
+    const events = this.getEvents(true, eventType, spell);
+    return events.length;
+  }
+
+  /**
+   * @param buffActive filters based on whether the buff is active or inactive (true for active, false for inactive)
+   * @param buff the spell object for the buff
+   * @param eventType the type of event that you want to search for. i.e. EventType.Cast, EventType.BeginCast, etc. Use EventType.Event for all events.
+   * @param cast an optional cast spell object to count. Omit or leave undefined to count all casts
+   * @returns the number of events found
+   */
+  countEventsByBuff(buffActive: boolean, buff: SpellInfo, eventType: EventType, cast?: SpellInfo) {
+    const events = buffActive
+      ? this.getEventsByBuff(true, buff, eventType, cast)
+      : this.getEventsByBuff(false, buff, eventType, cast);
+    return events.length;
+  }
+
+  /**
+   * @param buffActive filters based on whether the buff is active or inactive (true for active, false for inactive)
+   * @param buff the spell object for the buff
+   * @param eventType the type of event that you want to search for. i.e. EventType.Cast, EventType.BeginCast, etc. Use EventType.Event for all events.
+   * @param spell an optional spell object to search. Omit or leave undefined to count all events
+   * @returns an array of events that match the provided criteria
+   */
+  getEventsByBuff<ET extends EventType>(
+    buffActive: boolean,
+    buff: SpellInfo,
+    eventType: ET,
+    spell?: SpellInfo | SpellInfo[],
+  ): Array<MappedEvent<ET>> {
+    const events = this.getEvents(true, eventType, spell);
+    const filteredEvents = events.filter((e) =>
+      buffActive
+        ? this.selectedCombatant.hasBuff(buff.id, e.timestamp - 1)
+        : !this.selectedCombatant.hasBuff(buff.id, e.timestamp + 1),
+    );
+    return filteredEvents;
+  }
```

Comment:
I would turn this into two methods, tbh. `getEventsWithBuff` and `getEventsWithoutBuff`. slight code duplication, but it is easier to tell what is going on without having to consult the documentation for what the boolean parameter means.